Adds the tele-gulag [ready for review/merge]

lzimann opened the pull request:
Now security can send the bad guys to the gulag without any effort! 
Just put the unlucky guy inside the teleporter, insert a prisoner ID and hit the button! 
You don't even need to strip him, it's all automatic!

fix token #18417 

:cl: name: Lzimann
:add: Security now has the labor camp teleporter! Send those pesky greytiders to work with no effort!
/:cl:

lzimann committed:
Adds the tele-gulag

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+	var/obj/item/weapon/card/id/prisoner/id = null
+	var/obj/machinery/gulag_teleporter/teleporter = null
+	var/obj/structure/gulag_beacon/beacon = null
+	var/mob/living/carbon/human/prisoner = null
```
not necessary, machines have a ```occupant``` var.

lzimann replied:
are they always human though? (that's why i added this var)

RemieRichards replied:
No, but you just check where required and cast.

InsaneHyena commented:
Where is "FUCK" label?

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+		prisoner = teleporter.occupant
+		prisoner_list["name"] = prisoner.name
+		if(!isnull(data_core.general))
+			for(var/datum/data/record/R in data_core.security)
```
this should be an istypeless forloop.

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+/obj/machinery/computer/gulag_teleporter_computer/proc/findteleporter()
+	var/obj/machinery/gulag_teleporter/teleporterf = null
+
+	for(dir in list(NORTH,EAST,SOUTH,WEST))
```
this list already exists, it's called ```cardinal```

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+		if(!isnull(teleporterf) && teleporterf.is_operational())
+			return teleporterf
+
+	return null
```
this does nothing, remove it

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+	if(!isnull(beaconf))
+		return beaconf
+
+	return null
```
this does nothing, remove it

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+
+	return null
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/findbeacon()
```
this entire proc could be written as

``` return beaconf || locate(/obj/structure/gulag_beacon)```

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+	user << "<span class='notice'>You lean on the back of [src] and start pushing the door open... (this will take about [breakout_time] minutes.)</span>"
+	user.visible_message("<span class='italics'>You hear a metallic creaking from [src]!</span>")
+
+	if(do_after(user,(breakout_time*60*10), target = src)) //minutes * 60seconds * 10deciseconds
```
1*60*10...
so 600?

lzimann replied:
it's more to show the message to the user in minutes(i can change to keep it cleaner if that's a problem)

RemieRichards replied:
just change "[breakout_time] minutes" to "a minute" and make this 600

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+		return
+	toggle_open()
+
+/obj/machinery/gulag_teleporter/updateUsrDialog()
```
what are you doin

lzimann replied:
for some reason this was fucking up with open_machine/close_machine, doing this fixed that

RemieRichards replied:
I'll let this slide, the old system is deprecated and all you're doing here is turning it off for your type.

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+		ui.open()
+
+/obj/machinery/computer/gulag_teleporter_computer/ui_data(mob/user)
+	var/data = list()
```
```var/list/data = list()``` please, properly typed.

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+			var/new_goal = input("Set the amount of points:", "Points", id.goal) as num|null
+			if(!isnum(new_goal))
+				return
+			if(isnull(new_goal))
```
you have a lot of ```isnull()```s where you don't need them, eg:
```if(!new_goal)``` here

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+				return
+			if(isnull(new_goal))
+				new_goal = default_goal
+			id.goal = round(min(new_goal,1000)) //maximum 1000 points
```
what about negative goals?
```Clamp(new_goal, 0, 1000)```

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+
+	for(dir in list(NORTH,EAST,SOUTH,WEST))
+		teleporterf = locate(/obj/machinery/gulag_teleporter, get_step(src, dir))
+		if(!isnull(teleporterf) && teleporterf.is_operational())
```
```if(teleporterf && teleporterf.is_operational())```

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+	open_machine()
+
+/obj/machinery/gulag_teleporter/container_resist()
+	var/mob/living/user = usr
```
NO.

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+		user << "[src] is locked!"
+	open_machine()
+
+/obj/machinery/gulag_teleporter/container_resist()
```
```container_resist(mob/living/user)```

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+		if(!user || user.stat != CONSCIOUS || user.loc != src || state_open || !locked)
+			return
+
+		locked = 0
```
```locked``` should be using ```TRUE```/```FALSE``` not ```1```/```0```

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+		W.forceMove(get_turf(src))
+
+/obj/machinery/gulag_teleporter/proc/handle_prisoner(var/obj/item/id, var/datum/data/record/R)
+	if(!occupant)
```
this check isn't necessary, the below ```if(!ishuman(occupant))``` already covers it

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+		stored_items[prisoner] -= W
+		W.forceMove(get_turf(src))
+
+/obj/machinery/gulag_teleporter/proc/handle_prisoner(var/obj/item/id, var/datum/data/record/R)
```
no ```var/``` in proc arguments

RemieRichards commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,129 @@
+	//open/closed with no occupant
+	icon_state = initial(icon_state) + (state_open ? "_open" : "")
+
+/obj/machinery/gulag_teleporter/relaymove(mob/user as mob)
```
No ```as X``` in proc args

KorPhaeron commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,146 @@
+//computer that handle the points and teleports the prisoner
+/obj/machinery/computer/gulag_teleporter_computer
+	name = "Labor Camp teleporter console"
```
generally we dont capitalize these sorts of things

lzimann committed:
Fixes for the fix god.

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,140 @@
+	return ..()
+
+/obj/machinery/computer/gulag_teleporter_computer/proc/equip_id()
+	if(id && prisoner)
```
if as I said above, to remove the prisoner var, this would be ```if(id && ishuman(occupant))```

kevinz000 commented:
Ok we might need this in summer pop but normally this is overkill

PKPenguin321 commented:
security changes should cost 2 tokens

PKPenguin321 commented:
also i dont like this it'll just be used as a harder-to-escape permabrig

lzimann commented:
it's not perma though, you can't send someone without an ID and the max point is 1000.

Bawhoppen commented:
@PKPenguin321 ops changes should cost 3 aswell

Zilenan91 commented:
So a gib button? Because that's what this is.

Steelpoint commented:
All this change does is streamline the process of sending people to the gulag.

If your all that concerned then lobby to make the Gulag the new perma, or maybe DON'T commit a crime worthy of being thrown to the Gulag for a extended period.

Dorsisdwarf commented:
Except that the point of this change is to make the gulag timed sentences more common replacements, even though the gulag is now on Helldeath Murderland and prisoners are totally helpless against any enemies.

A 2 minute sentence for breaking a window can now = horrible death you will never, ever be retrieved from or cloned

Firecage commented:
Steelpoint. You can be send to the gulag for minor crimes, in the place on times sentences. It doesn't have to be major crimes. And as I said, for gulag prisoners, it is virtually a death sentence.

So I dislike this even though I agree with the general idea of this. Just a bit to late for this since its the wrong gulag.

Steelpoint commented:
Then we should upgrade the Gulag to the new perma and replace perma with a bunch of extra cells or something.

Firecage commented:
So basically force security to execute people instead of putting them in perma? No thanks.

Zilenan91 commented:
It's a gib button basically

Firecage commented:
@lzimann Well, if you also do work on to make Lavaland Gulag be less of a literal death sentence, then yah, then this PR would be fully awesome.

lzimann commented:
the biggest problem is mobs spawning there, they shouldn't, i'll look to fix it

swankcookie commented:
I've been saying this forever, Perma has been one of the worst implemented parts of security, ever. While i understand the concerns about the new Gulag being a little bit of a shitshow, at least the bad parts about it are, for the most part, unintended (mobs spawning where they should not, etc.).  Perma, on the other hand, is intentionally just a boring, slightly bigger, permanent cell that would work if TG had even a dash of RP left... which it does not. I think @Steelpoint had the best idea here. Security needs to be changed.

Dorsisdwarf commented:
It works quite well if 
A) More than 1 person goes in
and B) They're both willing to RP/shamelessly beg the admins for actual gardening kit

swankcookie commented:
unfortuanety, RP has become incredibly rare on TG.  Too the point where it might be dishonest to still refer to us as medRP.

lzimann commented:
We never got to the medrp point though

swankcookie commented:
ehhh. I'd say 2014 was relatively medRP

Bawhoppen commented:
bay is medRP, tg has always only been lightRP, however at times basil can be quite the RPzone

Bobylein commented:
But it's not even lightRP anymore, try as qm to get at least one form filled out if people ask for stuff, they will say: Fuck you! And come back with an ied to throw at your counter. 

Metagaming is the norm and every crewmember knows what a sleepy pen or a changeling is. 

And generally I love this teleporter idea as wardens usually don't even care to give out prisoner IDs but it will end in everyone who did a petty crime being teleported there.

tedward1337 commented:
Goon has the port-o-brig 
Why can't we have this

optimumtact commented:
another stale pr here to clog up my mind

optimumtact commented:
Also perma is literally worse than death

lzimann committed:
Tguis the point reclaimer console

lzimann committed:
Makes the machines constructable/deconstructabled
Added some missing sprites for the teleporter.

lzimann commented:
I'm going to fix the conflicts after I finish everything(almost there)

swankcookie commented:
@optimumtact Thank you

lzimann committed:
Finishes the teleporter and console
Also added a log when a person is teleported.

lzimann committed:
Hugboxes the lavaland gulag

KorPhaeron commented on code/modules/mining/mine_turfs.dm:
```diff
@@ -338,7 +338,7 @@
 /turf/open/floor/plating/asteroid/airless/cave/proc/SpawnMonster(turf/T)
 	if(prob(30))
-		if(istype(loc, /area/mine/explored))
+		if(istype(loc, /area/mine/explored) || istype(loc, /area/lavaland/surface/outdoors/labor_camp))
```
wrong way to do this

Just put low chance turfs on the gulag that dont have a chance to spawn caves

KorPhaeron replied:
Or make a new area lavaland/explored and use that instead of the gulag area since it'd make sense to reuse that around the base as well

KorPhaeron commented:
Making it low chance turfs would be better though otherwise you'll have the high lavaland drop rates with none of the risk

F-OS commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,155 @@
+		return
+	if(!allowed(usr))
+		usr << "<span class='warning'>You don't have the required access use it.</span>"
+		return
```
*to

F-OS commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,155 @@
+			teleporter.toggle_open()
+		if("teleporter_lock")
+			if(teleporter.state_open)
+				usr << "Close the teleporter before!"
```
before what?

xxalpha commented:
Because I am certain this will be abused by shit players :-1:

lzimann commented:
@xxalpha how so? people can already send folks to the gulag without an ID, or straight execute them, this will only make the process quicker but it requires an ID(with max 1000 points)

Zilenan91 commented:
If it's quicker, it's easier. This means it'll be far more attractive to send someone where they will never have any hope of rejoining the round.

Dorsisdwarf commented:
Just make the autogulag automatically do prison cards geez

lzimann commented:
How so @Zilenan91? Mobs spawning in the gulag area is not intented and IS being fixed here

Zilenan91 commented:
Not just mobs. The Gulag has always been effective. The only reason nobody uses it is because it's time consuming to equip them with everything they need. If this was added it'd just result in your round being over even worse than if they killed you since they just plug you in and there's not a chance in hell you'll ever get let out.

lzimann commented:
The gulag is already all automatic when you get sent with an ID. All you gotta do is mine the amount of points, go to the shuttle and send it back to the station...

Zilenan91 commented:
Yes, but sec has to get them the clothes, dress them, set their points manually, and then buckle them in and throw them out on the Gulag itself. This is a process that takes a good deal of time and is why most people don't use the Gulag

With the tele-gulag you just need to plug in an ID and then someone is out of the round forever

KorPhaeron commented:
I'm gonna merge this by the way

Dorsisdwarf commented:
>Yes, but sec has to get them the clothes, dress them, set their points manually, and then buckle them in and throw them out on the Gulag itself. This is a process that takes a good deal of time and is why most people don't use the Gulag

>With the tele-gulag you just need to plug in an ID and then someone is out of the round forever

*until they do the mining or get eaten by monsters

Zilenan91 commented:
1000 points take a real long time to mine. On old asteroid Gulag when I last got gulagged it took me around 35 minutes. This was before gulag prisoners had scanners and mesons though so it's likely different now but it's not an experience I want on anyone since it just consisted of me standing around hoping I didn't breach into an empty part of the asteroid and fucking atmos in the area since it had happened to me before and the Gulag had like 1 room that was atmos safe if atmos got fucked.

lzimann committed:
Adds the item reclaimer

lzimann committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into telegulag

Fix conflicts

bgobandit commented:
maybe if people weren't lazy non-RPing fucks they'd be able to leave the gulag

lzimann committed:
Map it in

lzimann commented:
Currently I mapped only on box. I'll do it on other maps later. I'll also update the OP later with tgui pics

lzimann committed:
Actually commit the item reclaimer

RemieRichards commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,159 @@
+	prisoner << "<span class='warning'>The teleportation makes you a little dizzy.</span>"
+	PoolOrNew(/obj/effect/particle_effect/sparks, prisoner.loc)
+	playsound(src.loc, "sparks", 50, 1)
+	if(teleporter.locked)
```
this isn't really necessary, you don't gain anything by only setting it FALSE if it's TRUE, it's probably actually SLOWER to check

RemieRichards commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,105 @@
+
+	var/list/mobs = list()
+	for(var/i in stored_items)
+		var/mob/thismob = i
```
so the keys to this list are mobs?

RemieRichards commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,105 @@
+	..()
+	addtimer(src, "locate_teleporter", 5)
+
+/obj/machinery/gulag_item_reclaimer/Destroy()
```
you haven't cleared out ```inserted_id``` or ```linked_teleporter```

RemieRichards commented:
@lzimann none of these machines have proper ```Destroy()```s

RemieRichards commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,105 @@
+	for(var/i in contents)
+		var/obj/item/I = i
+		I.forceMove(get_turf(src))
+	..()
```
All ```Destroy()```s must return ```..()```

tedward1337 commented:
Merge get

lzimann committed:
Fixes some stuff
Moves the reclaimer link to the teleporter
Adds a volcanic labor mineral rock turf

lzimann committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into telegulag

Fix conflicts

lzimann committed:
Adds a claim points button to the labor stacker

Armhulen commented:
isn't merged yet?

lzimann committed:
Fuck conflicts

optimumtact commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,164 @@
+
+/obj/machinery/computer/gulag_teleporter_computer/Destroy()
+	if(id)
+		id.loc = get_turf(src)
```
use force move please

optimumtact commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,164 @@
+		if(!id)
+			if(!user.drop_item())
+				return
+			W.loc = src
```
force move

optimumtact commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,164 @@
+					usr.put_in_hands(id)
+					id = null
+				else
+					id.loc = get_turf(src)
```
force move

optimumtact commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,164 @@
+				if(istype(I, /obj/item/weapon/card/id/prisoner))
+					if(!usr.drop_item())
+						return
+					I.loc = src
```
force move

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+	if(linked_teleporter)
+		linked_teleporter.linked_reclaimer = null
+	if(inserted_id)
+		inserted_id.loc = get_turf(src)
```
force move

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+					usr.put_in_hands(inserted_id)
+					inserted_id = null
+				else
+					inserted_id.loc = get_turf(src)
```
force move

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+				if(istype(I, /obj/item/weapon/card/id/prisoner))
+					if(!usr.drop_item())
+						return
+					I.loc = src
```
force move

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+		if(!inserted_id)
+			if(!user.drop_item())
+				return
+			I.loc = src
```
forceMove

optimumtact commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,173 @@
+	for(var/obj/item/W in occupant)
+		if(occupant.unEquip(W))
+			if(istype(W, /obj/item/weapon/restraints/handcuffs))
+				W.loc = get_turf(src)
```
forceMove

optimumtact commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,173 @@
+				linked_reclaimer.contents += W
+				W.loc = linked_reclaimer
+			else
+				W.loc = get_turf(src)
```
forceMove

optimumtact commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,173 @@
+			if(linked_reclaimer)
+				linked_reclaimer.stored_items[occupant] += W
+				linked_reclaimer.contents += W
+				W.loc = linked_reclaimer
```
forceMove

optimumtact commented on code/modules/mining/laborcamp/laborstacker.dm:
```diff
@@ -19,125 +19,119 @@
+		if(!inserted_id)
+			if(!user.drop_item())
+				return
+			I.loc = src
```
forceMove

optimumtact commented on code/modules/mining/laborcamp/laborstacker.dm:
```diff
@@ -19,125 +19,119 @@
+					usr.put_in_hands(inserted_id)
+					inserted_id = null
+				else
+					inserted_id.loc = get_turf(src)
```
forceMove

optimumtact commented on code/modules/mining/laborcamp/laborstacker.dm:
```diff
@@ -19,125 +19,119 @@
+				if(istype(I, /obj/item/weapon/card/id/prisoner))
+					if(!usr.drop_item())
+						return
+					I.loc = src
```
forceMove

optimumtact commented:
sorry :smile:

lzimann committed:
forcemove(oranges)

lzimann committed:
Last commit i swear

optimumtact commented on code/game/machinery/computer/gulag_teleporter.dm:
```diff
@@ -0,0 +1,164 @@
+					usr.put_in_hands(id)
+					id = null
+				else
+					id.forceMove(src)
```
I think you want to get_turf(src) first for these, forceMove doesn't handle that specifically, it handles a bunch of unrelated stuff around pulling items and ensuring crossed gets called.

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+	if(linked_teleporter)
+		linked_teleporter.linked_reclaimer = null
+	if(inserted_id)
+		inserted_id.forceMove(src)
```
get_turf again

optimumtact commented on code/game/machinery/gulag_item_reclaimer.dm:
```diff
@@ -0,0 +1,106 @@
+					usr.put_in_hands(inserted_id)
+					inserted_id = null
+				else
+					inserted_id.forceMove(src)
```
again probably want get_turf so it drops

optimumtact commented on code/game/machinery/gulag_teleporter.dm:
```diff
@@ -0,0 +1,173 @@
+	for(var/obj/item/W in occupant)
+		if(occupant.unEquip(W))
+			if(istype(W, /obj/item/weapon/restraints/handcuffs))
+				W.forceMove(src)
```
for the handcuffs, do you want to specifically put them on the floor instead so that sec officers can grab them back or is this intended behaviour were they get stored in the teleporters contents?

optimumtact commented on code/modules/mining/laborcamp/laborstacker.dm:
```diff
@@ -19,125 +19,119 @@
+					usr.put_in_hands(inserted_id)
+					inserted_id = null
+				else
+					inserted_id.forceMove(src)
```
get_turf again so the id gets dropped

optimumtact commented:
Apart from my  last few comments this is almost ready to go :+1: good work

lzimann committed:
Drops some stuff in the ground properly

lzimann committed:
Adds an emag function to item reclaimer

swankcookie commented:
woooo lets make security fun again!

lzimann committed:
Map the teleporter in MetaStation

optimumtact commented:
Did you properly map merge lavaland? it's showing a complete rewrite

lzimann commented:
What happened was: I made some changes and map merged normally, but the merge fucked up completly, then I replaced and merged, but conflicted and I don't know anymore what's going on

KorPhaeron commented:
its okay not like lavaland map matters

KorPhaeron commented:
gonna redo it next week anyway

lzimann commented:
all i changed was the gulag part, which now has the correct mineral spawner(which won't make tunnels) and the gulag beacon
